internal/builder/embedded/driver.go

```diff
@@ -0,0 +1,192 @@
+package embedded
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+
+	"github.com/containerd/containerd/namespaces"
+	"github.com/docker/distribution/reference"
+	controlapi "github.com/moby/buildkit/api/services/control"
+	"github.com/moby/buildkit/identity"
+	"github.com/moby/buildkit/session"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/dominodatalab/forge/internal/archive"
+	"github.com/dominodatalab/forge/internal/builder/config"
+	"github.com/dominodatalab/forge/internal/builder/embedded/bkimage"
+	"github.com/dominodatalab/forge/internal/builder/embedded/bkimage/types"
+)
+
+type driver struct {
+	bk               *bkimage.Client
+	contextExtractor archive.Extractor
+}
+
+func NewDriver() (*driver, error) {
+	client, err := bkimage.NewClient(getStateDir(), types.AutoBackend)
+	if err != nil {
+		return nil, fmt.Errorf("cannot create buildkit client: %w", err)
+	}
+
+	return &driver{
+		bk:               client,
+		contextExtractor: archive.FetchAndExtract,
+	}, nil
+}
+
+func (d *driver) BuildAndPush(ctx context.Context, opts *config.BuildOptions) ([]string, error) {
+	if len(opts.PushRegistries) == 0 {
+		return nil, errors.New("image builds require at least one push registry")
+	}
+
+	// HACK: configure registry hosts for every run and reset afterwards
+	original := d.bk.RegistryHosts
+	d.bk.RegistryHosts = generateRegistryHosts(opts.Registries)
+	defer func() {
+		d.bk.RegistryHosts = original
+	}()
```

Comment:
Wouldn't we need to change the underlying references since this is passed to in `createWorkerOpt`? I'm a little unclear why this hack is necessary / how it works.